### DISCUSSION TITLE
Plans of switch to UE5 Skeleton

### DISCUSSION THREAD

**Discussion by Solessfir** [Sixze/ALS-Refactored#91] [General]
Hi! Any plans/estimates of switching ALS to the new UE5 Skeleton? _// This is kinda duplicate question =)_
Or I can create a pull request to help with that

**Comment by Sixze** [comment]
In the future, I plan to use IK Rig to retarget animations from the old skeleton to the new one, and only then, if there are no issues, I may completely transfer ALS to the new skeleton.

**Reply by Solessfir** [reply]
Ok, thanks!

**Reply by murataka** [reply]
Will you start a branch so we can start contributing to it using manny ue5 skeleton ? Some components in marketplace began using ue5 skeleton by default , so using ue4 skeleton started bringing its own challeges already . The current status of Als seems quite stable and I feel it might be a good time to start it. Maybe we can retarget it properly .

**Comment by Donpoko** [comment]
Hi, any update on this ? I'm waiting for the UE5 update to switch from old ALS to this one.

**Comment by Sixze** [comment]
I've already released a new version of the plugin with support for Unreal Engine 5.0, but without the new skeleton yet. If you really want to use ALS with the new skeleton, I suggest you ask on the ALS Discord, they probably already made an IK Rig to retarget the animation from the old skeleton to the new one.

As for the integration of the new skeleton into the plugin, I can't say that it will be soon, unfortunately at the moment I don't have much free time to do this.

**Comment by AchimTuran** [comment]
I tried it as a Proof of Concept. With my TTToolbox based workflow I was able to add the required virtual bones, sockets, skeleton curves and blend masks/profiles. With some tinkering of the Linked Anim Instances, IK Retargeting does not set the matching target skeleton instances so we still have the AnimMan instances set. Settings those correctly gave me working locomotion motions.

As some bones (virtual and unweigthed (ik)) need to be constraint, we will have to rework the ik_hand_gun bone key frames. Additionally, the folder structure will get lost, which is normal if you retarget the animations...

Vanilla UE5.1 will mess up our anim curves and additive animations. Also the AdditiveAnimType is not correctly set and we might see intersecting hands with the body.

IKRetargeting in general currently not work very well within Unreal, okay it is since ages. UE4.X had similar issues.

Happily I fixed already anim curves, additive animations and correctly update AdditiveAnimType. AnimInstances might be fixed soon as well. Also I like to have a option where we can keep our folder structure! It is annyoing that this hard work gets lost...

UE5 Manny seems also to break the turn in place system. When looking at a different direction with the camera, the UE5 Manny did not automatically trigger the turn in place system.

The bone constraints I used are documented here:
https://github.com/Sixze/ALS-Refactored/discussions/15#discussioncomment-4382144

**Reply by AchimTuran** [reply]
Also I am still getting some warnings in the main AnimBP
![grafik](https://user-images.githubusercontent.com/6927011/207138293-e4b3f356-90c6-4367-97e8-dd4ea75ff0d4.png)

**Reply by AchimTuran** [reply]
Linked anim layers for overlay states also get not remapped.
![grafik](https://user-images.githubusercontent.com/6927011/207138676-4e8d4e2c-d47b-45b1-867a-055a2565d8df.png)

**Reply by AchimTuran** [reply]
Getting mantling working in a stable way was not successful.
I had to reimplement this function and duplicate + adapt the data assets where the montages are set.
![grafik](https://user-images.githubusercontent.com/6927011/207141782-78a5276c-337e-4217-8d98-6685abc4193e.png)
![grafik](https://user-images.githubusercontent.com/6927011/207142258-1ed1710e-93a5-4e4c-88b2-2b5f52dcb63f.png)


Then somehow the montages where sometimes played, sometimes not.

As I am only doing a Proof of Concept here, I will now stop fixing the issues. It is defenitely possible but requires some effort, not a out of the box working solution exist. My main goal was to have a setup to see how the locomotion animations of ALS can be changed. ;-)

**Reply by tornellihenrique** [reply]
Hello!
I managed to achieve a very good result retargeting ALS-Refactored, your TTToolbox was kinda useful on this process so I'd like to notify you.
Feel free to check out [my comment](https://github.com/Sixze/ALS-Refactored/discussions/91#discussioncomment-7719806) on this discussion.

**Comment by murataka** [comment]
This thread is kind of stuck because of retargeting breaks things , right ? Otherwise we could retartget and replace references , and it would work out of the box ?

**Reply by faxlab** [reply]
Dude Ive got no idea about what you're asking, but if I replace the ALS character skeletal mesh to Manny from ue5, it works kinda ok, but the character animations are super fast, its almost funny. Not sure if there is a fix but if someone knows, please reply and share!

**Reply by Benbentwo** [reply]
make sure your animations have the gait and pose curves

**Comment by tornellihenrique** [comment]
Hello! I know this is kinda old, but I have some interesting progress on retargeting ALS-Refactored to the UE5 Manny.
Unfortunately the only missing stuff is related to the overlays anims, the base locomotion seems to be good.

I'm learning Maya so that I can fix these anims directly into it, I'm afraid the UE retargeter will never be enough for this IK part (specially for the ik_hand_gun retargeting) and also for the skeleton size difference.

Either way, the retargeting I managed to do is one of the best that you may find around the internet for now. If anyone that knows Blender/Maya wants to help me finish this work, I'd highly appreciate it, feel free to contact me on discord: tornelli (Tornelli#4688)

Some important notes before cloning:
- I'm constantly updating this fork with the original repo (this one)
    - I'm trying to follow any modifications to AB_Als and CR_Als so that I can fix on the retargeted one.
    - I'm using the Als C++ classes, so you don't need to worry about changes on them.
- I've made two retarget methods
    - LiveRetarget: 10% finished, I'm planning to take a look at [this live retarget method](https://www.youtube.com/watch?v=_-DIqFJvZyw) soon: 
    - FullRetarget: 80% finished, I've retarget all the Linked Anim Graphs (haven't retarget the Monolithic YET).
- All the mantling, transitions, getup, etc. is working almost perfectly.
- The issue with the overlays:
    - The ik_hand_gun bone still needs to be manually adjusted for all the overlays poses.
    - If you use the sockets (activated by default on this fork) you might reach closer to a good result (I didn't adjusted them properly yet).
    - The bow overlay anims needs to be almost completly reworked using some Anim Tool.
- Final thoughts:
    - I'm studying Maya when possible so that I can make this Manny version the perfect ALS version.
    - I'll be manually adjusting every single animation so that them work perfectly with ALS logic. (this may take some good time if no one helps).

Fork: https://github.com/Ruskivotka/ALS-Refactored-Manny

Thank you and good development :)
